src/providers/idp/oidc_child_handler.c

```diff
@@ -0,0 +1,426 @@
+
+    io->pid = pid;
+
+    /* Set file descriptors. */
+    io->read_from_child_fd = pipefd_from_child[0];
+    io->write_to_child_fd = pipefd_to_child[1];
+    PIPE_FD_CLOSE(pipefd_from_child[1]);
+    PIPE_FD_CLOSE(pipefd_to_child[0]);
+    sss_fd_nonblocking(io->read_from_child_fd);
+    sss_fd_nonblocking(io->write_to_child_fd);
+
+    /* Setup the child handler. It will free io and remove it from the hash
+     * table when it exits. */
+    ret = child_handler_setup(ev, pid, child_exited, io, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Could not set up child signal handler "
+              "[%d]: %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    /* Steal the io pair so it can outlive this request if needed. */
+    talloc_steal(idp_req->idp_options, io);
+
+    *_child_pid = pid;
+    *_io = io;
+
+    ret = EOK;
+
+done:
+    if (ret != EOK) {
+        PIPE_CLOSE(pipefd_from_child);
+        PIPE_CLOSE(pipefd_to_child);
+        child_terminate(pid);
+    }
+
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t create_send_buffer(struct idp_req *idp_req,
+                                  struct io_buffer **io_buf)
+{
+    struct io_buffer *buf = NULL;
+    const char *client_secret;
+    int ret;
+
+    buf = talloc_zero(idp_req, struct io_buffer);
+    if (buf == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc failed.\n");
+        return ENOMEM;
+    }
+
+    client_secret = dp_opt_get_cstring(idp_req->idp_options,
+                                       IDP_CLIENT_SECRET);
+    if (client_secret == NULL || *client_secret == '\0') {
+        ret = EOK;
+        goto done;
+    }
+
+    buf->size = strlen(client_secret);
+    buf->data = talloc_size(buf, buf->size);
+    if (buf->data == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_size failed.\n");
+        talloc_free(buf);
+        return ENOMEM;
+    }
+
+    safealign_memcpy(buf->data, client_secret, strlen(client_secret), NULL);
+
+    ret = EOK;
+
+done:
+    if (ret == EOK) {
+        *io_buf = buf;
+    }
+
+    return ret;
+}
+
+static void handle_oidc_child_send_done(struct tevent_req *subreq);
+static void handle_oidc_child_done(struct tevent_req *subreq);
+
+struct tevent_req *handle_oidc_child_send(TALLOC_CTX *mem_ctx,
+                                         struct tevent_context *ev,
+                                         struct idp_req *idp_req,
+                                         struct io_buffer *send_buffer)
+{
+    struct tevent_req *req, *subreq;
+    struct handle_oidc_child_state *state;
+    int ret;
+    struct io_buffer *buf = NULL;
+
+    req = tevent_req_create(mem_ctx, &state, struct handle_oidc_child_state);
+    if (req == NULL) {
+        return NULL;
+    }
+
+    state->ev = ev;
+    state->idp_req = idp_req;
+    state->buf = NULL;
+    state->len = 0;
+    state->child_pid = -1;
+    state->timeout_handler = NULL;
+
+    if (send_buffer == NULL) {
```

Comment:
Can this really happen?
Both 'auth' and 'id' create 'send_buffer' before calling `handle_oidc_child_send()`.

Reply:
I  changed the 'id' code to not create the buffer because the 'id' code only has to send the client secret and nothing else so that it can be added here.